classes/migration/upgrade/v3_4_0/I7191_EditorAssignments.php

```diff
@@ -0,0 +1,145 @@
+<?php
+
+/**
+ * @file classes/migration/upgrade/v3_4_0/I7191_EditorAssignments.inc.php
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
+ * Copyright (c) 2000-2021 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class I7191_EditorAssignments
+ * @brief Update the subeditor_submission_group table to accomodate new editor assignment settings
+ */
+
+namespace PKP\migration\upgrade\v3_4_0;
+
+use Exception;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Schema;
+use PKP\install\DowngradeNotSupportedException;
+
+abstract class I7191_EditorAssignments extends \PKP\migration\Migration
+{
+    /**
+     * Adds a user_group_id column to the subeditor_submission_group
+     * table and adds initial data. Adds foreign keys where appropriate.
+     */
+    public function up(): void
+    {
+        if (empty($this->sectionDb) || empty($this->sectionIdColumn) || empty($this->contextColumn)) {
+            throw new Exception('Upgrade could not be completed because required properties for the I7191_EditorAssignments migration are undefined.');
+        }
+
+        Schema::table('subeditor_submission_group', function (Blueprint $table) {
+            $table->bigInteger('user_group_id')->nullable();
+        });
+
+        $this->setUserGroup();
+        $this->deleteOrphanedAssignments();
+        $this->addAutomatedAssignments();
+
+        Schema::table('subeditor_submission_group', function (Blueprint $table) {
+            $table->bigInteger('user_group_id')->nullable(false)->change();
+            $table->foreign('user_id')->references('user_id')->on('users');
+            $table->foreign('user_group_id')->references('user_group_id')->on('user_groups');
+        });
+    }
+
+    /**
+     * Reverse the downgrades
+     *
+     * @throws DowngradeNotSupportedException
+     */
+    public function down(): void
+    {
+        Schema::table('subeditor_submission_group', function (Blueprint $table) {
+            $table->dropForeign('user_group_id');
+            $table->dropColumn('user_group_id');
+        });
+    }
+
+    /**
+     * Set the user group for all existing assignments
+     *
+     * Uses the default user group for Role::ROLE_ID_SUB_EDITOR
+     * if one exists
+     */
+    protected function setUserGroup(): void
+    {
+        DB::table('user_groups')
```

Comment:
Maybe I've missed something, but I think that this will fall back as you suggest. The `orderBy('is_default')` is used to put the default user group first, but shouldn't exclude non-default user groups, right?

```php
DB::table('user_groups')
    ->where('role_id', '=', 17) // Role::ROLE_ID_SUB_EDITOR
    ->orderBy('is_default')
    ->get(['user_group_id', 'context_id'])
    ->each(function ($row) {
        DB::table('subeditor_submission_group')
            ->whereNull('user_group_id')
            ->where('context_id', '=', $row->context_id)
            ->update(['user_group_id' => $row->user_group_id]);
    });
```

Although now that I look at it, I think `->each()` should be `->first()`, right? That would get me the first user group in the set, with a default user group ordered before others?

> If there are several assumptions being made, maybe we ought to revisit e.g. being able to remove default user groups?

I would like to do this but maybe only once we have reduced the number of default user groups? The list feels built for a large publishing operation at the moment, and I can see lots of good UX reasons to delete the defaults for a small journal.

Reply:
Right on all counts, I think. It should be `first()`, and let's come back to defaults later.